roadmap/implementers-guide/src/node/approval/approval-voting.md

```diff
@@ -0,0 +1,247 @@
+    * If the cert kind is `RelayVRFDelay`, then we check if the VRF is valid for the validator's key with the input `block_entry.relay_vrf_story ++ cert.core_index.encode()` as described in [the approvals protocol section](../../protocol-approval.md#assignment-criteria). The cert can be ignored if the block did not cause inclusion of a candidate on that core index. Otherwise, this is a valid assignment for the included candidate. The delay tranche for the assignment is determined by reducing `(vrf.make_bytes().to_u64() % (session_info.n_delay_tranches + session_info.zeroth_delay_tranche_width)).saturating_sub(session_info.zeroth_delay_tranche_width)`.
+    * `import_checked_assignment`
+    * return the appropriate `VoteCheckResult` on the response channel.
+
+#### `ApprovalVotingMessage::CheckAndImportApproval`
+
+On receiving a `CheckAndImportApproval(indirect_approval_vote, response_channel)` message:
+  * Fetch the `BlockEntry` from the indirect approval vote's `block_hash`. If none, return `VoteCheckResult::Bad`.
+  * Fetch the `CandidateEntry` from the indirect approval vote's `candidate_index`. If the block did not trigger inclusion of enough candidates, return `VoteCheckResult::Bad`.
+  * Construct a `SignedApprovalVote` using the candidate hash and check against the validator's approval key, based on the session info of the block. If invalid or no such validator, return `VoteCheckResult::Bad`.
+  * Send `VoteCheckResult::Accepted`,
+  * `import_checked_approval(BlockEntry, CandidateEntry, ValidatorIndex)`
+
+#### `ApprovalVotingMessage::ApprovedAncestor`
+
+On receiving an `ApprovedAncestor(Hash, BlockNumber, response_channel)`:
+  * Iterate over the ancestry of the hash all the way back to block number given, starting from the provided block hash.
+  * Keep track of an `all_approved_max: Option<Hash>`.
+  * For each block hash encountered, load the `BlockEntry` associated. If any are not found, return `None` on the response channel and conclude.
+  * If the block entry's `approval_bitfield` has all bits set to 1 and `all_approved_max == None`, set `all_approved_max = Some(current_hash)`.
+  * If the block entry's `approval_bitfield` has any 0 bits, set `all_approved_max = None`.
+  * After iterating all ancestry, return `all_approved_max`.
+
+### Utility
+
+#### `import_checked_assignment`
+  * Load the candidate in question and access the `approval_entry` for the block hash the cert references.
+  * Ensure the validator index is not part of the backing group for the candidate.
```

Comment:
I suppose backers vote with their original backing announcement message since we do not distinguish the first backer.